### ISSUE TITLE
Recognize Stable annotation in JIT

### ISSUE THREAD

**Issue by gita-omr** [eclipse-openj9/openj9#12710]
Recognize Stable annotation in JIT using helpers implemented in [1]

[1] https://github.com/eclipse-openj9/openj9/pull/12679

**Comment by gita-omr** [comment]
**Design**

External behavior of the Stable annotation is described in [1]. It's important to add that all elements of multidimensional arrays should be treated as stable if the whole array was annotated as stable.

To recognize stable fields, we will add this into VMJ9.h:

   `virtual bool canDereferenceAtCompileTimeIfNotNull(TR::SymbolReference *fieldRef,  TR::Compilation *comp);`

which will be used by 

`J9::TransformUtil::transformIndirectLoadChainImpl()
`
similar to the way it uses 

`fej9->canDereferenceAtCompileTime(symRef, comp);`

except that just before folding it will check if the current value is not NULL.

This will handle stable fields but not array elements.  Constant array elements are already handled by `tryFoldCompileTimeLoad` in  OMR VPHandlers.cpp: https://github.com/eclipse/omr/blob/39f680adff633008210ef52d8b12549be598160f/compiler/optimizer/VPHandlers.cpp#L400

However, we need to distinguish those from stable elements (that can have two values). I propose to create a new type of shadow:

`StableArrayShadow()` and use it similar to` ImmutableArrayShadow()` mainly just to be recognized by `transformIndirectLoadChainImpl`. This will  handle elements of one dimensional arrays. 

In order to handle multi-dimensional arrays, I think, we need to create a new type of constraints and keep propagating it from one dimension to another.

[1] https://github.com/eclipse-openj9/openj9/issues/11990#issue-810315209

**Comment by gita-omr** [comment]
@jdmpapin @vijaysun-omr @0xdaryl Please let me know what you think.

@tajila fyi

**Comment by jdmpapin** [comment]
I dealt with a situation similar to stable fields in #12745: with `LambdaForm`, `CallSite.target` can be folded, but only sometimes. I think the structure I put in place to deal with that is general enough to deal with stable fields as well (except for the contents of stable arrays, which will take more doing). I handle it like this:
- `canDereferenceAtCompileTime()` means that it might be possible to fold the field, rejecting any field that we definitely can't fold
- `avoidFoldingInstanceField()` looks at a field (which must satisfy `canDereferenceAtCompileTimeWithFieldSymbol()`) and the particular object we want to load from, and determines whether to inhibit folding

In the doc comment for `avoidFoldingInstanceField()`, I even suggested folding only non-null values as a possible use case (though I had no idea about the stable annotation)

**Comment by gita-omr** [comment]
@jdmpapin thanks a lot for the comments! A field can be both final and stable. We need to attempt dereferencing  if field is final (or equivalent to that) or stable. Then, if the field is NOT final and has value 0, or  is CallSite.target, we should not dereference. So I think we need a method that tells us a field is final or equivalent to final (always can be dereferenced).

**Comment by jdmpapin** [comment]
The problem with `CallSite.target` is that it *is* final, so final can't mean always foldable

**Comment by gita-omr** [comment]
Sure, then we can exclude it from the "always can be folded method". My point is that we can't include Stable into it.

**Comment by gita-omr** [comment]
Could we make foldFinalFieldsIn() return false for CallSite.target?

**Comment by gita-omr** [comment]
I think the problem is that there is no such thing as "definitely-foldable" until we start folding, due to the CallSite.target :)

**Comment by gita-omr** [comment]
I would propose:
1) put the body of `canDereferenceAtCompileTime() `into something like `behavesAsFinal()` (we can discuss the name)
2) replace the body of  `canDereferenceAtCompileTime()`  with `return behavesAsFinal() || isStable()`
3) inside` AvoidInstanceField` add another check and return true if `(!behavesAsFinal() and the value is null)`
4) the actual folding (vs dereferencing)happens in `transformIndirectLoadChainImpl`.  There, we also need to check `(!behavesAsFinal() and the value is null)`. In fact, I think we don't need to check for NULL inside `dereferenceStructPointerChain` since we are not going to dereference NULL anyway.

So basically, I think both our PRs can coexist. Mine will check during actual folding and yours during dereferencing.

**Comment by jdmpapin** [comment]
That sounds pretty similar to what I was suggesting. A few clarifications though:
- [The check in (3) is the same as the "additional" check in (4)](https://github.com/eclipse-openj9/openj9/blob/4feda5f0a1010425f491ff7928412c610d757423/runtime/compiler/optimizer/J9TransformUtil.cpp#L451), since in general it's not until the latter point that we know which object we're loading the field from
- The check for null is still necessary when folding in `transformIndirectLoadChainImpl()`, because the null could be the result of the outermost load, in which case that load would normally be folded

Also, I think we can easily get away without `behavesAsFinal()` if we arrange things like this:
- return true for stable fields in `canDereferenceAtCompileTime()` (really `can...WithFieldSymbol()`)
- in (3), check `isStable()` and then only look for null when `isStable() && !isFinal()`.

`isFinal()` will differ from `behavesAsFinal()` for a few fields, but not for any stable fields. They'll only really differ for a few recognized fields in the OpenJ9 java/lang/invoke package, which we're in the process of replacing. And that's our code anyway, so I don't know why we didn't, or couldn't now, just make those fields actually final... In any case though, they're not also annotated stable.

(Technically there are also fields satisfying `isFinal()` in classes that aren't to be trusted, [but we already know we're not dealing with one of those](https://github.com/eclipse-openj9/openj9/blob/4feda5f0a1010425f491ff7928412c610d757423/runtime/compiler/optimizer/J9TransformUtil.cpp#L206-L209).)

IMO the code would be easier to understand at a high level without the `behavesAsFinal()` query. And in (3), I think that the null checking logic in `avoidFoldingInstanceField()` would be clearer this way, since it's explicitly restricted to stable fields. In your (3) as written, that check is already restricted to stable fields, but the reader has to reason
```
   (canDereferenceAtCompileTime() && !behavesAsFinal())
== ((behavesAsFinal() || isStable()) && !behavesAsFinal())
== isStable()
```

**Comment by gita-omr** [comment]
> `isFinal()` will differ from `behavesAsFinal()` for a few fields, but not for any stable fields. They'll only really differ for a few recognized fields in the OpenJ9 java/lang/invoke package, which we're in the process of replacing. And that's our code anyway, so I don't know why we didn't, or couldn't now, just make those fields actually final... In any case though, they're not also annotated stable.

Yes, with this assumption, I would agree with your approach.
